controllers/zookeeperbackup_controller.go

```diff
@@ -0,0 +1,403 @@
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+	changed := zookeeperBackup.WithDefaults()
+	if changed {
+		r.Log.Info("Setting default settings for zookeeper-backup")
+		if err := r.Client.Update(context.TODO(), zookeeperBackup); err != nil {
+			return reconcile.Result{}, err
+		}
+		return reconcile.Result{Requeue: true}, nil
+	}
+
+	// Check if zookeeper cluster exists
+	foundZookeeperCluster := &zookeeperv1beta1.ZookeeperCluster{}
+	zkCluster := zookeeperBackup.Spec.ZookeeperCluster
+	err = r.Client.Get(context.TODO(), types.NamespacedName{Name: zkCluster, Namespace: zookeeperBackup.Namespace}, foundZookeeperCluster)
+	if err != nil && errors.IsNotFound(err) {
+		r.Log.Error(err, fmt.Sprintf("Zookeeper cluster '%s' not found", zkCluster))
+		return reconcile.Result{}, err
+	}
+
+	// Define a new PVC object
+	pvc := newPVCForZookeeperBackup(zookeeperBackup)
+	// Set ZookeeperBackup instance as the owner and controller
+	if err := controllerutil.SetControllerReference(zookeeperBackup, pvc, r.Scheme); err != nil {
+		return reconcile.Result{}, err
+	}
+	if pvc.Annotations == nil {
+		pvc.Annotations = make(map[string]string)
+	}
+
+	// Calculate hash of PVC Spec
+	hash, err = hashstructure.Hash(pvc.Spec, hashstructure.FormatV2, nil)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	pvcHashStr := strconv.FormatUint(hash, 10)
+
+	// Check if PVC already created
+	foundPVC := &corev1.PersistentVolumeClaim{}
+	err = r.Client.Get(context.TODO(), types.NamespacedName{Name: pvc.Name, Namespace: pvc.Namespace}, foundPVC)
+	if err != nil && errors.IsNotFound(err) {
+		r.Log.Info("Creating a new PersistenVolumeClaim")
+		pvc.Annotations["last-applied-hash"] = pvcHashStr
+		err = r.Client.Create(context.TODO(), pvc)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+	} else if err != nil {
+		return reconcile.Result{}, err
+	} else {
+		// Check if pvc requires to be updated
+		if foundPVC.Annotations["last-applied-hash"] == pvcHashStr {
+			r.Log.Info("PVC already exists and looks updated", "pvc.Namespace", foundPVC.Namespace, "pvc.Name", foundPVC.Name)
+		} else {
+			pvc.Annotations["last-applied-hash"] = pvcHashStr
+			r.Log.Info("Update PVC", "Namespace", pvc.Namespace, "Name", pvc.Name)
+			err = r.Client.Update(context.TODO(), pvc)
+			if err != nil {
+				r.Log.Error(err, "PVC cannot be updated")
+				return reconcile.Result{}, err
+			}
+		}
+	}
+
+	// Define a new CronJob object
+	cronJob := newCronJobForCR(zookeeperBackup)
+	// Set ZookeeperBackup instance as the owner and controller
+	if err := controllerutil.SetControllerReference(zookeeperBackup, cronJob, r.Scheme); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Check Zookeeper Cluster status
+	if foundZookeeperCluster.Status.Replicas != foundZookeeperCluster.Status.ReadyReplicas {
```

Comment:
It might be desirable to compare the ReadyReplicas to `Spec.Replicas` to consider the possibility of situation when some of the replicas have not been created yet.